### PR TITLE
script to wrap unitary modifications by composite modification

### PR DESCRIPTION
**Pull request by ghazwarhili**

### CODE REVIEW

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+with
+    new_modification as (insert into networkmodifications.modification(id, date, group_id, modifications_order, stashed, message_type, message_values, type)
+        values (gen_random_uuid(), NULL, NULL, NULL, false, NULL, NULL, 'COMPOSITE_MODIFICATION') returning id),
+    new_composite as (insert into networkmodifications.composite_modification(id) (select id from new_modification) returning id),
+    modifications_ids as (select id from directory.element where type = 'MODIFICATION')
+insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+select n.id as id, m.id as modification_id, row_number() over () from modifications_ids m full join new_composite n on true=true;
+delete from directory.element where type = 'MODIFICATION';
+insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values ((select id from networkmodifications.composite_modification), 'Generated-modification-composite', 'admin',
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql]: Why 'admin' ? The owner should not change

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        '054e7336-c9cc-4dd9-9817-7a8b38d2a667', 'MODIFICATION', null, '2024-07-02 14:24:50.296498 +00:00',
+        '2024-07-01 21:14:15.517000 +00:00', 'admin', null, false);
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:16]: Insert with previous date. Or use a function like now()

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    new_composite as (insert into networkmodifications.composite_modification(id) (select id from new_modification) returning id),
+    modifications_ids as (select id from directory.element where type = 'MODIFICATION')
+insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+select n.id as id, m.id as modification_id, row_number() over () from modifications_ids m full join new_composite n on true=true;
+delete from directory.element where type = 'MODIFICATION';
+new_directory as (insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values (gen_random_uuid(), 'Modification_composite_migration', 'admin',
+        null, 'DIRECTORY', null, '2024-07-02 14:24:50.296498 +00:00', '2024-07-02 14:24:50.296498 +00:00', 'admin', null, false) return id);
+insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values (new_composite, 'Generated_modification_composite', 'admin',
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:14]: why admin ? The owner should not change

Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:14]: I think we should keep the previous name. 
Or something like "Composite modification of 'previous name'"

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 4]
```diff
@@ -7,8 +7,11 @@ with
 insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
 select n.id as id, m.id as modification_id, row_number() over () from modifications_ids m full join new_composite n on true=true;
 delete from directory.element where type = 'MODIFICATION';
+new_directory as (insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values (gen_random_uuid(), 'Modification_composite_migration', 'admin',
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:11]: The script should be as much transparent as possible for users !
Do not create a new create. Modifications should stay in the directory they are

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    new_modification as (insert into networkmodifications.modification(id, date, group_id, modifications_order, stashed, message_type, message_values, type)
+        values (gen_random_uuid(), NULL, NULL, NULL, false, NULL, NULL, 'COMPOSITE_MODIFICATION') returning id),
+    new_composite as (insert into networkmodifications.composite_modification(id) (select id from new_modification) returning id),
+    modifications_ids as (select id from directory.element where type = 'MODIFICATION')
+insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+select n.id as id, m.id as modification_id, row_number() over () from modifications_ids m full join new_composite n on true=true;
+delete from directory.element where type = 'MODIFICATION';
+new_directory as (insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values (gen_random_uuid(), 'Modification_composite_migration', 'admin',
+        null, 'DIRECTORY', null, '2024-07-02 14:24:50.296498 +00:00', '2024-07-02 14:24:50.296498 +00:00', 'admin', null, false) return id);
```
Comment (Tristan-WorkGH) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:12]:
```suggestion
        null, 'DIRECTORY', null, now(), now(), 'admin', null, false) return id);
```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
+values (new_composite, 'Generated_modification_composite', 'admin',
+        new_directory, 'MODIFICATION', null, '2024-07-02 14:24:50.296498 +00:00',
+        '2024-07-01 21:14:15.517000 +00:00', 'admin', null, false);
```
Comment (Tristan-WorkGH) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:16]:
```suggestion
        new_directory, 'MODIFICATION', null, now(), now(), 'admin', null, false);
```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+commit;
```
Comment (Tristan-WorkGH) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:17]: Missing EOL
```suggestion
commit;

```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql (hunk 5 above)
Comment (Tristan-WorkGH) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:12]:
```suggestion
    new_directory as (insert into directory.element(id, name, owner, parent_id, type, description, creation_date, last_modification_date, last_modified_by, stash_date, stashed)
values (gen_random_uuid(), 'Modification_composite_migration', 'admin', null, 'DIRECTORY', null, now(), now(), 'admin', null, false) returning id)
```

And you're not in the `with` scope here ...  
> <statement> expected, got 'new_directory'
screenshot

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]
Script not working as is...

**Review by Tristan-WorkGH — APPROVED** [review]
why so many empty lines?

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    new_modification as (
+
+        insert into networkmodifications.modification(id, date, group_id, modifications_order, stashed, message_type, message_values, type)
+
+            select m.new_id, NULL, NULL, NULL, false, NULL, NULL, 'COMPOSITE_MODIFICATION' from dir_element_with_m_id m
+
+            returning id
+    ),
+
+    new_composite as (insert into networkmodifications.composite_modification(id) select id from new_modification returning id),
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:16]:
```suggestion
    new_composite_modification_id as (insert into networkmodifications.composite_modification(id) select id from new_modification returning id),
```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+    new_element_modification as (
+
+        insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+
+            select m.new_id, m.id, row_number() over () as modifications_order from dir_element_with_m_id m
+
+            left join new_composite nc on m.new_id=nc.id
+
+            returning id, modification_id
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:26]:
```suggestion
            returning id
```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            returning id
+    ),
+
+    new_composite as (insert into networkmodifications.composite_modification(id) select id from new_modification returning id),
+
+    new_element_modification as (
+
+        insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+
+            select m.new_id, m.id, row_number() over () as modifications_order from dir_element_with_m_id m
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:22]:
```suggestion
            select m.new_id, m.id, 0 as modifications_order from dir_element_with_m_id m
```

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 11]
```diff
@@ -0,0 +1,31 @@
+begin transaction isolation level serializable;
+
+with
+
+    dir_element_with_m_id as (select *, gen_random_uuid() as new_id from directory.element where type = 'MODIFICATION'),
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:5]: Can't we only work with this common table expression and avoid all the other "as".
Actually, we only need to work with oldId and newId, they don't change

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        insert into networkmodifications.modification(id, date, group_id, modifications_order, stashed, message_type, message_values, type)
+
+            select m.new_id, NULL, NULL, NULL, false, NULL, NULL, 'COMPOSITE_MODIFICATION' from dir_element_with_m_id m
+
+            returning id
+    ),
+
+    new_composite as (insert into networkmodifications.composite_modification(id) select id from new_modification returning id),
+
+    new_element_modification as (
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:18]: new_element_modification_id

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: migration/migrate_modifications_in_directory_to_composite_modification_fix.sql [hunk 13]
```diff
@@ -0,0 +1,12 @@
+begin transaction isolation level serializable;
+update networkmodifications.public.modification e
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification_fix.sql:2]:
```suggestion
update networkmodifications.modification e
```

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: migration/migrate_modifications_in_directory_to_composite_modification.sql [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+update directory.element e set id = (select modifs.new_id from MODIFICATION_ELEMENTS modifs where modifs.id=e.id) where type = 'MODIFICATION';
+
+insert into networkmodifications.modification (id, date, group_id, modifications_order, stashed, message_type, message_values, type)
+select new_id, current_date, NULL, 0, false, 'COMPOSITE_MODIFICATION', '{}', 'COMPOSITE_MODIFICATION' from MODIFICATION_ELEMENTS;
+
+insert into networkmodifications.composite_modification (id) (select new_id from MODIFICATION_ELEMENTS);
+
+insert into networkmodifications.composite_modification_sub_modifications(id, modification_id, modifications_order)
+    (select new_id, id, 0 from MODIFICATION_ELEMENTS);
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification.sql:12]: Add : 
drop table MODIFICATION_ELEMENTS;

File: migration/migrate_modifications_in_directory_to_composite_modification_fix.sql [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+set date = CURRENT_DATE,
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification_fix.sql:3]:
```suggestion
set date = (select creation_date from modification_elements where id = e.id) ,
```

File: migration/migrate_modifications_in_directory_to_composite_modification_fix.sql [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+    modifications_order = 0,
+    message_type = 'COMPOSITE_MODIFICATION',
+    message_values = '{}'
+where type = 'COMPOSITE_MODIFICATION'
+  AND date IS NULL
+  AND modifications_order IS NULL
+  AND message_type IS NULL
+  AND message_values IS NULL;
```
Comment (etiennehomer) [inline migration/migrate_modifications_in_directory_to_composite_modification_fix.sql:11]: add : 
drop table MODIFICATION_ELEMENTS;

**Review by etiennehomer — CHANGES_REQUESTED** [review]

**Review by etiennehomer — APPROVED** [review]

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**